src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -0,0 +1,337 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Net.Quic;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal
+{
+    internal class QuicStreamContext : TransportConnection, IQuicStreamFeature
+    {
+        private Task _processingTask;
+        private QuicStream _stream;
+        private QuicConnectionContext _connection;
+        private readonly QuicTransportContext _context;
+        private readonly CancellationTokenSource _streamClosedTokenSource = new CancellationTokenSource();
+        private IQuicTrace _log;
+        private string _connectionId;
+        private const int MinAllocBufferSize = 4096;
+        private bool _streamClosed;
+        private volatile Exception _shutdownReason;
+        private readonly TaskCompletionSource<object> _waitForConnectionClosedTcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+        private readonly object _shutdownLock = new object();
+
+        public QuicStreamContext(QuicStream stream, QuicConnectionContext connection, QuicTransportContext context)
+        {
+            _stream = stream;
+            _connection = connection;
+            _context = context;
+            _log = context.Log;
+
+            ConnectionClosed = _streamClosedTokenSource.Token;
+
+            var maxReadBufferSize = context.Options.MaxReadBufferSize.Value;
+            var maxWriteBufferSize = context.Options.MaxWriteBufferSize.Value;
+
+            // TODO should we allow these PipeScheduler to be configurable here?
+            var inputOptions = new PipeOptions(MemoryPool, PipeScheduler.ThreadPool, PipeScheduler.Inline, maxReadBufferSize, maxReadBufferSize / 2, useSynchronizationContext: false);
+            var outputOptions = new PipeOptions(MemoryPool, PipeScheduler.Inline, PipeScheduler.ThreadPool, maxWriteBufferSize, maxWriteBufferSize / 2, useSynchronizationContext: false);
+
+            var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
+
+            Features.Set<IQuicStreamFeature>(this);
+
+            // TODO populate the ITlsConnectionFeature (requires client certs).
+            Features.Set<ITlsConnectionFeature>(new FakeTlsConnectionFeature());
+            IsUnidirectional = !stream.CanWrite;
```

Comment:
What if it's a write-only stream? Isn't that what would happen if StartUnidirectionalStreamAsync was called, right? It makes me wonder if we shouldn't copy QuicStream and do CanRead/CanWrite instead of IsUnidirectional.